oap-server/oal-rt/src/main/java/org/apache/skywalking/oal/rt/OALRuntime.java

```diff
@@ -194,7 +196,13 @@ private void generateClassAtRuntime(OALScripts oalScripts) throws OALCompileExce
         }
 
         for (Map.Entry<String, DispatcherContext> entry : allDispatcherContext.getAllContext().entrySet()) {
-            dispatcherClasses.add(generateDispatcherClass(entry.getKey(), entry.getValue()));
+            final String fullClassName = dispatcherClassName(entry.getKey(), true);
+            if (DISPATCHER_CLASS_CACHE.containsKey(fullClassName)) {
+                continue;
+            }
+            final Class dispatcherClass = generateDispatcherClass(entry.getKey(), entry.getValue());
+            dispatcherClasses.add(dispatcherClass);
+            DISPATCHER_CLASS_CACHE.put(fullClassName, dispatcherClass);
```

Comment:
> When could regenerate happen?

Because we have tcp-related metrics rules in a separate `.oal` file, and need to load that file into OALRuntime, when this happens, the loader will try to define dispatcher classes for those entities in `tcp.oal`, which contains `Service`, `ServiceInstance`, etc., they exist in `core.oal` too, hence will be defined twice

Reply:
I am not sure whether this is correct. About the dispatcher, it includes the `filter` logic. Once you skip those, I have a feeling this logic will be lost.

Reply:
I think you should use OAL definition to separate the package names of the same ServiceDispatcher? Does this work?

Reply:
> I am not sure whether this is correct. About the dispatcher, it includes the filter logic. Once you skip those, I have a feeling this logic will be lost.

Yes I missed that.

> I think you should use OAL definition to separate the package names of the same ServiceDispatcher? Does this work?

Yes, have enhanced the class to allow customized dispatcher class package